pkg/googlegroups/groupreader.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package googlegroups
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	admin "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/cloudidentity/v1"
+
+	"github.com/abcxyz/pkg/logging"
+	"github.com/abcxyz/team-link/pkg/groupsync"
+)
+
+const (
+	MemberTypeUser  = "USER"
+	MemberTypeGroup = "GROUP"
+)
+
+// Ensure we conform to the interface.
+var _ groupsync.GroupReader = (*GroupReader)(nil)
+
+// GroupReader provides read operations for groups and users in GCP.
+type GroupReader struct {
+	identity *cloudidentity.Service
+	admin    *admin.Service
+}
+
+// NewGroupReader create a new GroupReader.
+func NewGroupReader(identityService *cloudidentity.Service, adminService *admin.Service) *GroupReader {
+	return &GroupReader{
+		identity: identityService,
+		admin:    adminService,
+	}
+}
+
+// Descendants retrieve all users (children, recursively) of a group.
+func (g GroupReader) Descendants(ctx context.Context, groupID string) ([]*groupsync.User, error) {
+	var members []*groupsync.User
+	err := g.identity.Groups.Memberships.SearchTransitiveMemberships(groupID).Context(ctx).Pages(ctx,
+		func(page *cloudidentity.SearchTransitiveMembershipsResponse) error {
+			for _, m := range page.Memberships {
+				// we only want user memberships but the API doesn't give us any type information.
+				// Instead, we infer the type from the resource name. e.g. groups have a resource
+				// name of the form `groups/%s` and users of the form 'users/%d'
+				if strings.HasPrefix(m.Member, "users/") {
+					members = append(members, &groupsync.User{ID: m.Member})
+				}
+			}
+			return nil
+		},
+	)
+	return members, err //nolint:wrapcheck
```

Comment:
Optional: how much do we really trust the GCP identity API to give us a friendly error message here? We might consider:

 ```
if err != nil {
  return nil, fmt.Errorf("SearchTransitiveMemberships: %w", err)
}
```

The situation I'm trying to avoid is the annoying case where the service fails with an unhelpful message like "a thing was not found" and we have no idea where it came from. Maybe that doesn't apply here.

Reply:
Good suggestion, I've wrapped the error in with a message.